modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/internal/HttpHeadersAuthenticatorUtils.java

```diff
@@ -0,0 +1,125 @@
+                    // forwarding the request beyond the headers part
+                    listener.onResponse(null);
+                }, e -> listener.onFailure(new HttpHeadersValidationException(e))));
+            } else {
+                // cannot authenticate the request because it's not wrapped correctly, see {@link #wrapAsMessageWithAuthenticationContext}
+                listener.onFailure(new HttpHeadersValidationException(new IllegalStateException("Cannot authenticate unwrapped requests")));
+            }
+        }, threadContext);
+    }
+
+    /**
+     * Given a {@link DefaultHttpRequest} argument, this returns a new {@link DefaultHttpRequest} instance that's identical to the
+     * passed-in one, but the headers of the latter can be authenticated, in the sense that the channel handlers returned by
+     * {@link #getValidatorInboundHandler(HttpValidator, ThreadContext)} can use this to convey the authentication result context.
+     */
+    public static HttpMessage wrapAsMessageWithAuthenticationContext(HttpMessage newlyDecodedMessage) {
+        assert newlyDecodedMessage instanceof HttpRequest;
+        DefaultHttpRequest httpRequest = (DefaultHttpRequest) newlyDecodedMessage;
+        HttpHeadersWithAuthenticationContext httpHeadersWithAuthenticationContext = new HttpHeadersWithAuthenticationContext(
+            newlyDecodedMessage.headers()
+        );
+        return new DefaultHttpRequest(
+            httpRequest.protocolVersion(),
+            httpRequest.method(),
+            httpRequest.uri(),
+            httpHeadersWithAuthenticationContext
+        );
+    }
+
+    /**
+     * Returns the authentication thread context for the {@param request}.
+     */
+    public static ThreadContext.StoredContext extractAuthenticationContext(org.elasticsearch.http.HttpRequest request) {
+        HttpHeadersWithAuthenticationContext authenticatedHeaders = unwrapAuthenticatedHeaders(request);
+        return authenticatedHeaders != null ? authenticatedHeaders.authenticationContextSetOnce.get() : null;
+    }
+
+    /**
+     * Translates the netty request internal type to a {@link HttpPreRequest} instance that code outside the network plugin has access to.
+     */
+    public static HttpPreRequest asHttpPreRequest(HttpRequest request) {
+        return new HttpPreRequest() {
+
+            @Override
+            public RestRequest.Method method() {
+                return translateRequestMethod(request.method());
+            }
+
+            @Override
+            public String uri() {
+                return request.uri();
+            }
+
+            @Override
+            public Map<String, List<String>> getHeaders() {
+                return getHttpHeadersAsMap(request.headers());
+            }
+        };
+    }
+
+    private static HttpHeadersWithAuthenticationContext unwrapAuthenticatedHeaders(org.elasticsearch.http.HttpRequest request) {
+        if (request instanceof HttpPipelinedRequest) {
```

Comment:
Can you help me understand why we don't need to do this in 8.x ? https://github.com/elastic/elasticsearch/blob/main/modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/internal/HttpHeadersAuthenticatorUtils.java#L110

Reply:
In 7.17 there is an extra netty pipeline handler https://github.com/elastic/elasticsearch/blob/fe18a67f025d645daf9399d84bb01c57768ea18a/modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java#L336 that assigns the received id tag to requests, by wrapping them in this `HttpPipelinedRequest` type https://github.com/elastic/elasticsearch/blob/fe18a67f025d645daf9399d84bb01c57768ea18a/modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java#L47 .
In main, there's no wrapping. The pipelining (HTTP proto) and request dispatch happen in the same channel handler https://github.com/elastic/elasticsearch/blob/991b06e23504ecdcf82cfaad96952160bd53dd6a/modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java#L114 .

Hence the code in the `HttpHeadersAuthenticatorUtils` has to peel away one extra layer in 7.17 to get to the type that contains the authentication.